### PR TITLE
feat: Binary cross entropy loss added to funtional experimental

### LINKED ISSUE #23910: binary_cross_entropy

### PR DESCRIPTION
**Pull request by RummanAli**
# Binary Cross Entropy Loss added to experimental.

## Related Issue

Close #23910

### CODE REVIEW

**Comment by RummanAli** [comment]
@mohame54 1 test is failing due to some mismatch in outputs of paddle backend and tensorflow frontend
image

**Comment by mohame54** [comment]
Hi @RummanAli thanks for your contribution,

**Comment by mohame54** [comment]
I have some questions and modifications.

File: ivy/data_classes/array/experimental/losses.py
Comment (mohame54) [inline ivy/data_classes/array/experimental/losses.py:1]: you should modify your doc string a little bit like here
```
"""
        ivy.Array instance method variant of ivy.binary_cross_entropy. This method simply wraps the
        function, and so the docstring for ivy.binary_cross_entropy also applies to this method with
        minimal changes.
        Parameters
        ----------
        self
            input array containing true labels. # you should here specify what the input array is like you could specify that they  are  probabilities.
        target
            input array containing targeted labels. # also here you should specify that they are the same shape as input array with values between 0 and 1.
"""
```

File: ivy/data_classes/container/experimental/losses.py
Comment (mohame54) [inline ivy/data_classes/container/experimental/losses.py:1]: Also in here.

File: ivy/functional/backends/jax/experimental/losses.py
Comment (mohame54) [inline ivy/functional/backends/jax/experimental/losses.py:1]: - I don't know why you wrote this function **poisson_nll_loss**  and didn't use it.
- also you wrote **_apply_loss_reduction** and didn't use it.

File: ivy/functional/backends/numpy/experimental/losses.py
Comment (mohame54) [inline ivy/functional/backends/numpy/experimental/losses.py:1]: also in here you didn't use **_apply_loss_reduction** function

File: ivy/functional/backends/jax/experimental/losses.py
Comment (RummanAli) [reply ivy/functional/backends/jax/experimental/losses.py:1]: So basically poison loss function was added to main branch and it had conflicts with my branch so I just merged the upstream branch to my code.

**Comment by mohame54** [comment]
also you should look into lint errors and resolve them and follow the semantic check. @RummanAli

File: ivy/functional/backends/numpy/experimental/losses.py
Comment (RummanAli) [reply ivy/functional/backends/numpy/experimental/losses.py:1]: This was also from upstream main branch

File: ivy/functional/ivy/experimental/losses.py
[the remark this answers is not in the document]
Comment (RummanAli) [reply ivy/functional/ivy/experimental/losses.py:1]: okok

**Comment by RummanAli** [comment]
> also you should look into lint errors and resolve them and follow the semantic check. @RummanAli

okkk

File: ivy/functional/backends/jax/experimental/losses.py
Comment (mohame54) [reply ivy/functional/backends/jax/experimental/losses.py:1]: Oh, if that's the case

**Comment by mohame54** [comment]
Hi @RummanAli could resolve your PR conflicts and check my suggestions? thank you.

**Comment by RummanAli** [comment]
> Hello, thanks for reviewing. I have made all the changes as you asked.

**Comment by RummanAli** [comment]
> Hi @RummanAli could resolve your PR conflicts and check my suggestions? thank you.

**Comment by mohame54** [comment]
> @mohame54 1 test is failing due to some mismatch in outputs. the error says: "the results from backend paddle and ground truth framework tensorflow do not match" image
 
Your implementation for each backend should give the same outputs with little `unnoticeable` differences so if that error appeared to you, you should check your backend implementations.

**Comment by RummanAli** [comment]
> > @mohame54 1 test is failing due to some mismatch in outputs. the error says: "the results from backend paddle and ground truth framework tensorflow do not match" image
> 
> @RummanAli Your implementation for each backend should give the same outputs with little `unnoticeable` differences so if that error appeared to you, you should check your backend implementations.

Yea I kind of resolved it but currently anew error has propped up:
image
do you know what could be the issue here?

**Comment by RummanAli** [comment]
@mohame54 and tests are not running due to this error.

**Comment by mohame54** [comment]
Hi @RummanAli everything seems ok but I have some requests if you could fulfill them I will merge your PR I know it's too much but bear with me 😄.

- [ ] you should add the weight argument to **binary_cross_entropy**  function to fulfill the superset behavior like in [here](https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html)  
- [ ] you should use **_apply_loss_reduction** in your backend functions to avoid code repetitions.
- [ ] you should solve your code lint errors

**Comment by mohame54** [comment]
> @mohame54 and tests are not running due to this error.

you could change the ground truth backend to another backend for example `jax` or `torch`

**Comment by mohame54** [comment]
> > > @mohame54 1 test is failing due to some mismatch in outputs. the error says: "the results from backend paddle and ground truth framework tensorflow do not match" image
> > 
> > 
> > @RummanAli Your implementation for each backend should give the same outputs with little `unnoticeable` differences so if that error appeared to you, you should check your backend implementations.
> 
> Yea I kind of resolved it but currently anew error has propped up: image do you know what could be the issue here?
could you show me a pic of the whole error msg cuz I can't see anything from the above pic 😆

**Comment by RummanAli** [comment]
> Hi @RummanAli everything seems ok but I have some requests if you could fulfill them I will merge your PR I know it's too much but bear with me 😄.
> 
>     * [ ]  you should add the weight argument to **binary_cross_entropy**  function to fulfill the superset behavior like in [here](https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html)
> 
>     * [ ]  you should use **_apply_loss_reduction** in your backend functions to avoid code repetitions.
> 
>     * [ ]  you should solve your code lint errors

Okkk sure, thanks for reviewing. Let me fix and add the thing you mentioned 🖖

**Comment by RummanAli** [comment]
image
image
image
@mohame54 this is the whole traceback

**Comment by mohame54** [comment]
Hmm, It seems that the error is not related to your code could you try to run another existing test in test_losses file to check if this problem appears only to you or it's more general.

**Comment by RummanAli** [comment]
yea the same error appears with kl_div, but this error vanishes as soon as I revert back to previous commit.

**Comment by RummanAli** [comment]
@mohame54  I have made the changes as you requested and about _apply_loss_reduction that is from the upstream branch as well which I had to merge to resolve conflicts.

**Comment by RummanAli** [comment]
Happy to change anything else if required!

**Comment by mohame54** [comment]
Hi @RummanAli, I see you added weight argument but I have some questions suppose we have the probabilities shape of
 (N, L, C) and the target should be also of the same shape (N, L, C) and you have a weight of shape (N,) wouldn't it cause mismatching errors

**Comment by mohame54** [comment]
also you should solve your lint errors 
image
here as you can see some doc string lines should be shortened

File: ivy/functional/backends/numpy/experimental/losses.py [hunk 1]
```diff
@@ -164,3 +164,24 @@ def poisson_nll_loss(
...
+    target: np.ndarray,
+    /,
+    *,
+    weight: Optional[np.ndarray] = None,
+    reduction: Optional[str] = "mean",
+    out: Optional[np.ndarray] = None,
+) -> np.ndarray:
+    loss = -1 * weight * (target * np.log(input) + (1 - target) * np.log(1 - input))
+
+    if reduction == "mean":
```
Comment (mohame54) [inline ivy/functional/backends/numpy/experimental/losses.py:182]: You should use _apply_loss_reduction function instead of handling it yourself

File: ivy/functional/backends/tensorflow/experimental/losses.py [hunk 2]
```diff
@@ -156,3 +156,26 @@ def poisson_nll_loss(
...
+    reduction: Optional[str] = "mean",
+    out: Optional[tf.Tensor] = None,
+) -> tf.Tensor:
+    loss = (
+        -1
+        * weight
+        * (target * tf.math.log(input) + (1 - target) * tf.math.log(1 - input))
+    )
+
+    if reduction == "mean":
```
Comment (mohame54) [inline ivy/functional/backends/tensorflow/experimental/losses.py:177]: also in here

File: ivy/functional/backends/tensorflow/experimental/losses.py [hunk 3]
```diff
@@ -156,3 +156,26 @@ def poisson_nll_loss(
...
+
+@with_unsupported_dtypes({"2.13.0 and below": ("bool", "bfloat16")}, backend_version)
+def binary_cross_entropy(
+    input: tf.Tensor,
+    target: tf.Tensor,
+    /,
+    *,
+    weight: Optional[tf.Tensor] = None,
+    reduction: Optional[str] = "mean",
+    out: Optional[tf.Tensor] = None,
```
Comment (mohame54) [inline ivy/functional/backends/tensorflow/experimental/losses.py:169]: you should move the out argument from your function arguments

**Review by mohame54 — CHANGES_REQUESTED** [review]

**Comment by mohame54** [comment]
you should also solve your test error I think there is something wrong related to the decorators you are using @RummanAli

**Comment by RummanAli** [comment]
> Hi @RummanAli, I see you added weight argument but I have some questions suppose we have the probabilities shape of (N, L, C) and the target should be also of the same shape (N, L, C) and you have a weight of shape (N,) wouldn't it cause mismatching errors

I just checked the implementation on colab and it works well. So basicall the weights, input and target are multiplied element-wise:
image

**Comment by mohame54** [comment]
> > Hi @RummanAli, I see you added weight argument but I have some questions suppose we have the probabilities shape of (N, L, C) and the target should be also of the same shape (N, L, C) and you have a weight of shape (N,) wouldn't it cause mismatching errors
> 
> I just checked the implementation on colab and it works well. So basicall the weights, input and target are multiplied element-wise: image

Did you check for every backend like tensorflow or jax ?

**Comment by mohame54** [comment]
that's weird but Okay.

**Comment by mohame54** [comment]
still remains to check why there are errors in your tests and the modification of reduction part in numpy, tensorflow, jax backends and some lint errors

**Comment by RummanAli** [comment]
> still remains to check why there are errors in your tests and the modification of reduction part in numpy, tensorflow, jax backends and some lint errors

I just pushed the changes for lint code errors and loss reduction is being used now. I am just unable to figure out why the tests are not running. The tests are not running for any of the loss function in the file but they run as soon as I revert back to previous commit: but that code has conflicts with main branch.

**Comment by mohame54** [comment]
> > still remains to check why there are errors in your tests and the modification of reduction part in numpy, tensorflow, jax backends and some lint errors
> 
> I just pushed the changes for lint code errors and loss reduction is being used now. I am just unable to figure out why the tests are not running. The tests are not running for any of the loss function in the file but they run as soon as I revert back .to previous commit: but that code has conflicts with main branch.

that's weird 😕

**Comment by RummanAli** [comment]
> > > still remains to check why there are errors in your tests and the modification of reduction part in numpy, tensorflow, jax backends and some lint errors
> > 
> > 
> > I just pushed the changes for lint code errors and loss reduction is being used now. I am just unable to figure out why the tests are not running. The tests are not running for any of the loss function in the file but they run as soon as I revert back .to previous commit: but that code has conflicts with main branch.
> 
> that's weird 😕

I fixed this error, there was an conflict of distribution due to numpy as it seems. But I am getting this error now:
image
This is shrinker file.

**Comment by RummanAli** [comment]
image

**Comment by RummanAli** [comment]
While all the tests are passing for poisson_nll_loss: Which signifies there is something wrong with my code 🤔

**Comment by RummanAli** [comment]
fixed that currently getting this error:
image

**Comment by mohame54** [comment]
you should check them.

**Comment by RummanAli** [comment]
> you should check them.

image
finally all tests cases are passing.

**Comment by mohame54** [comment]
Hi @RummanAli thanks for your PR it looks good to me.
